src/tracing/metric_macros.h

```diff
@@ -0,0 +1,177 @@
+        ::chip::Tracing::Internal::LogEvent(_metric_event);                                                                        \
+    } while (false)
+
+/**
+ *  @def MATTER_LOG_METRIC
+ *
+ *  @brief
+ *    When tracing is enabled, this macro generates a metric event and logs it to the tracing backend.
+ *
+ *  Example usage:
+ *  @code
+ *      MATTER_LOG_METRIC(PASESession, err);
+ *  @endcode
+ *      The above example generates an instant metric event with key kMetricPASESession.
+ *      The metric also holds the 32 bit value corresponding to the ChipError object 'err'.
+ *
+ *  @param[in]  key The key representing the metric name/event. The name is one the kMetricXYZ string literal
+ *                  constants as defined in metric_event.h.
+ *
+ *  @param[in]  value An optional value for the metric. This value corresponds to one of the values supported
+ *                    in MetricEvent::Value
+ */
+#define MATTER_LOG_METRIC(key, ...) __MATTER_LOG_METRIC(Tag::Instant, key, ##__VA_ARGS__)
+
+/**
+ * @def MATTER_TRACE_BEGIN
+ *
+ * @brief
+ * Generate a metric with the Begin tag
+ *
+ *  Example usage:
+ *  @code
+ *      MATTER_LOG_METRIC_BEGIN(PASESession);
+ *  @endcode
+ *      The above example generates a Begin metric event with key kMetricPASESession.
+ *
+ *  @param[in]  key The key representing the metric name/event. The name is one the kMetricXYZ string literal
+ *                  constants as defined in metric_event.h.
+ *
+ *  @param[in]  value An optional value for the metric. This value corresponds to one of the values supported
+ *                    in MetricEvent::Value
+ */
+#define MATTER_LOG_METRIC_BEGIN(key, ...) __MATTER_LOG_METRIC(Tag::Begin, key, ##__VA_ARGS__)
+
+/**
+ * @def MATTER_TRACE_END
+ *
+ * @brief
+ * Generate a metric with the Eng tag
+ *
+ *  Example usage:
+ *  @code
+ *      MATTER_LOG_METRIC_END(PASESession);
+ *  @endcode
+ *      The above example generates an End metric event with key kMetricPASESession.
+ *
+ *  @param[in]  key The key representing the metric name/event. The name is one the kMetricXYZ string literal
+ *                  constants as defined in metric_event.h.
+ *
+ *  @param[in]  value An optional value for the metric. This value corresponds to one of the values supported
+ *                    in MetricEvent::Value
+ */
+#define MATTER_LOG_METRIC_END(key, ...) __MATTER_LOG_METRIC(Tag::End, key, ##__VA_ARGS__)
+
+#else // Tracing is disabled
+
+#define __MATTER_LOG_METRIC_DISABLE(...)                                                                                           \
+    do                                                                                                                             \
+    {                                                                                                                              \
+    } while (false)
+
+// Noop for logging metrics
+#define MATTER_LOG_METRIC(...) __MATTER_LOG_METRIC_DISABLE(__VA_ARGS__)
```

Comment:
I think only these should be scoped somewhere else.

in particular it seems these should be in `macros.h` and only the tracing extraction macros should be in a separate header.

Reply:
I have this all now in a metric_macros.h

Reply:
The base building blocks should be in macros.h ... we do not have `begin_end_macros.h` and `instant_macros.h` but rather all ase macros go into one single header that does the ifdef.

The `WithMetrics` can be split into a separate `return_macros` (I believe that sounds more clear than `metric_macros.h`. 

My intent would be to have "metrics are a feature of tracing" and not completely separate it out.

Reply:
If you want metrics to be incorporated in the tracing macros and this just above using the TRACE as the name, we can do that but the existing TRACE macros are not enough as they stand. The START and END only expect strings, which is not flexible enough.